### ISSUE TITLE
Dangling tray icon after Tray is being released.

### ISSUE THREAD

**Issue by pronebird** [electron/electron#8597]
* Electron version: 1.5.0
* Operating system: macOS

I have a simple component that creates Tray once mounted and destroys it when unmounted. However, it does not happen if I reload page manually. It simply creates new tray icon and leaks the other one.

```js
  componentWillMount() {
    const iconPath = path.join(__dirname, '../assets/trayicon.png');
    this.tray = new remote.Tray(iconPath);
  }

  componentWillUnmount() {
    this.tray.destroy();
    this.tray = null;
  }
```

### Expected behavior

Instances of `remote.Tray` should automatically call `destroy` when released.

### Actual behavior

Tray icon keeps dangling in menubar when there is no reference left to Tray instance.

### How to reproduce

Create an instance of Tray in renderer process. Refresh the page.

**Comment by MarshallOfSound** [comment]
@pronebird Is `componentWillUnmount` actually being called on unload/reload?  I seem to remember having issues with this before where it is actually a race condition as to whether componentWillUnmount is actually fired?

**Comment by groundwater** [comment]
Hey all! I am assuming `componentWillUnmount` is part of a React app? Is there a repro case for this that doesn't involve react, or is the problem that React and electron just won't play well together?

**Comment by pronebird** [comment]
@groundwater It's React yes. You can easily reproduce this by creating Tray menu from renderer process. Since entire web view is being trashed on page reload, the reference to tray menu is being lost too. There is probably a way to create it in backend process which is never reloaded and somehow share the reference with renderer.

**Comment by MarshallOfSound** [comment]
>  But then over time, detached tray menus are being released and removed from menubar automatically. So it leaves me wonder how that works.

I believe that is just periodic GC you are seeing there

I think one thing I want to convey to people is just because the `remote` module is there, it's isn't a lifeline to be used for everything 😆.  I think your best bet with this is either.

* Don't rely on `componentWillUnmount`, manually attach a `beforeunload` listener is `componentDidMount` and use that to destroy your tray icon.
* Don't create the tray icon through the `remote` module in a renderer process, instead create it in the main process by listening to IPC events you can fire from the renderer process.

**Comment by zeke** [comment]
Any luck getting it working, @pronebird?

**Comment by pronebird** [comment]
@zeke I moved tray creation in main process. I haven't tried passing it to renderer via IPC since I changed my mind about having popup menu in tray, but as @MarshallOfSound mentioned it's possible to pass tray object via IPC.

I haven't tried but something like that should work:

```js
// main
const tray = new Tray(__dirname + '/money.png');
ipcMain.on('give me tray', (e) => e.returnValue = tray);

// renderer
const tray = ipcRenderer.sendSync('give me tray');
// do something with tray...
```

Closing this since there is not much can be done about when and how GC collects garbage.
